core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateLeftUnnestWithGlobalAggregation.java

```diff
@@ -0,0 +1,277 @@
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.optimizations.PlanNodeSearcher;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.AssignUniqueId;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.planner.plan.CorrelatedJoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.planner.plan.UnnestNode;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.matching.Pattern.nonEmpty;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.sql.planner.iterative.rule.Util.restrictOutputs;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static io.trino.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static io.trino.sql.planner.plan.AggregationNode.singleGroupingSet;
+import static io.trino.sql.planner.plan.JoinNode.Type.LEFT;
+import static io.trino.sql.planner.plan.Patterns.CorrelatedJoin.correlation;
+import static io.trino.sql.planner.plan.Patterns.CorrelatedJoin.filter;
+import static io.trino.sql.planner.plan.Patterns.correlatedJoin;
+import static io.trino.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * This rule finds correlated UnnestNode in CorrelatedJoinNode's subquery and folds
+ * them into UnnestNode representing LEFT JOIN UNNEST.
+ * This rule transforms plans, where:
+ * - UnnestNode in subquery is based only on correlation symbols
+ * - UnnestNode in subquery is LEFT without filter
+ * - subquery contains global aggregation over the result of unnest
+ * Additionally, other global aggregations, grouped aggregations and projections
+ * in subquery are supported.
+ * <p>
+ * Transforms:
+ * <pre>
+ * - CorrelatedJoin (LEFT or INNER) on true, correlation(c1, c2)
+ *      - Input (a, c1, c2)
+ *      - Aggregation
+ *           global grouping
+ *           agg <- agg1(x)
+ *           - Projection
+ *                x <- foo(y)
+ *                - Aggregation
+ *                     group by (g)
+ *                     y <- agg2(u)
+ *                     - Unnest LEFT
+ *                          g <- unnest(c1)
+ *                          u <- unnest(c2)
+ *                          replicate: ()
+ * </pre>
+ * Into:
+ * <pre>
+ * - Projection (restrict outputs)
+ *      - Aggregation
+ *           group by (a, c1, c2, unique)
+ *           agg <- agg1(x)
+ *           - Projection
+ *                x <- foo(y)
+ *                - Aggregation
+ *                     group by (g, a, c1, c2, unique)
+ *                     y <- agg2(u)
+ *                     - Unnest LEFT
+ *                          g <- unnest(c1)
+ *                          u <- unnest(c2)
+ *                          replicate: (a, c1, c2, unique)
```

Comment:
why do we need c1, c2 here?

Reply:
Oh - we need those. As they are also part of `CorrelatedJoinNode` which was not obvious to me. 
Probably can not be part of grouping set. But maybe it does not matter.

Reply:
NVM: https://github.com/trinodb/trino/pull/6326/files#r580313959

Sorry for thinking out loud :)